app/src/main/java/io/neurolab/main/TestModeActivity.java

```diff
@@ -91,77 +95,114 @@ public void onReceive(Context context, Intent intent) {
         setContentView(R.layout.activity_test_mode);
-        // setting up the UsbManager instance with the desired USB service.
-        usbManager = (UsbManager) getSystemService(this.USB_SERVICE);
-        // initializing the DeviceConnector instance for checking and connecting to the device.
-        deviceConnector = new DeviceConnector(usbManager);
-        // to start reading from the Arduino.
+
+        // Setting listeners of all the three buttons
+        beginBtn = findViewById(R.id.buttonBegin);
         readBtn = findViewById(R.id.buttonRead);
-        // to send and write data to the Arduino.
         writeBtn = findViewById(R.id.buttonWrite);
-        // to input the data to be sent.
+        stopBtn = findViewById(R.id.buttonStop);
+
+        // to start reading from the Arduino.
         editText = findViewById(R.id.editText);
-        // to display the data sent and received through the serial port.
+        // to display the data received through the serial port.
         displayView = findViewById(R.id.displayData);
         // to choose a baud rate.
+
+        beginBtn.setOnClickListener(v -> searchForArduinoDevice());
+        writeBtn.setOnClickListener(v -> write(v));
+        stopBtn.setOnClickListener(v -> closeConnection());
+
         baudRateSpinner = findViewById(R.id.baud_rates_spinner);
         ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this, R.array.baud_rates_array, android.R.layout.simple_spinner_item);
         adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
-        // setting the adapter with the spinner.
         baudRateSpinner.setAdapter(adapter);
         baudRateSpinner.setOnItemSelectedListener(this);
-        writeBtn.setEnabled(false);
+
+        // setting up the UsbManager instance with the desired USB service.
+        usbManager = (UsbManager) getSystemService(this.USB_SERVICE);
+        // initializing the DeviceConnector instance for checking and connecting to the device.
+        deviceConnector = new DeviceConnector(usbManager);
+
         IntentFilter intentFilter = new IntentFilter();
         // adding the possible USB intent actions.
         intentFilter.addAction(ACTION_USB_PERMISSION);
-        intentFilter.addAction(UsbManager.ACTION_USB_ACCESSORY_ATTACHED);
-        intentFilter.addAction(UsbManager.ACTION_USB_ACCESSORY_DETACHED);
         registerReceiver(broadcastReceiver, intentFilter);
     }
 
-    /**
-     * To start the connection and reading from the Arduino.
-     *
-     * @param view
+    /*
+        Searches for all connected devices and then check if the vendor ID of the Arduino
+        matches that of a connected device.
+        If found, permission must be requested from the user.
      */
-    public void beginRead(View view) {
-        deviceConnector.checkConnection(this);
-        device = deviceConnector.getDevice();
+    private void searchForArduinoDevice(){
```

Comment:
@suresh-gandhi A similar function exists in DeviceConnector class for the same purpose. If you want this function here, then you can remove it from the other place or use it from there. Minimizes redundancy.

Reply:
@jddeep we will actually convert it into a utility class and then implement the singleton pattern to make it accessible everywhere in the app. I don't think we should do that for two classes and therefore I kept the code in a single class. Maybe for some helper functions, it may work but this is not a helper function.

Reply:
Yes. Please remove the same function from the DeviceConnector as this ain't a helper method.